tests/datasets/test_entity_factory.py

```diff
@@ -0,0 +1,77 @@
+import threading
+from typing import Any
+
+from snuba.datasets.entities import EntityKey
+from snuba.datasets.entities.factory import (
+    get_all_entity_names,
+    get_entity,
+    get_entity_name,
+)
+from snuba.datasets.entity import Entity
+
+ENTITY_KEYS = [
+    EntityKey.DISCOVER,
+    EntityKey.EVENTS,
+    EntityKey.GROUPASSIGNEE,
+    EntityKey.GROUPEDMESSAGES,
+    EntityKey.OUTCOMES,
+    EntityKey.OUTCOMES_RAW,
+    EntityKey.SESSIONS,
+    EntityKey.ORG_SESSIONS,
+    EntityKey.TRANSACTIONS,
+    EntityKey.DISCOVER_TRANSACTIONS,
+    EntityKey.DISCOVER_EVENTS,
+    EntityKey.METRICS_SETS,
+    EntityKey.METRICS_COUNTERS,
+    EntityKey.ORG_METRICS_COUNTERS,
+    EntityKey.METRICS_DISTRIBUTIONS,
+    EntityKey.PROFILES,
+    EntityKey.FUNCTIONS,
+    EntityKey.REPLAYS,
+    EntityKey.GENERIC_METRICS_SETS,
+    EntityKey.GENERIC_METRICS_DISTRIBUTIONS,
+]
+
+# This test should be the first to ensure dataset factory module is has fresh set of objects
```

Comment:
It occurs to me that you should be able to reset the value of `_ENT_FACTORY` in the test to ensure that it gets rebuilt.

```
from snuba.datasets.entities import factory # You can't import _ENT_FACTORY directly since that will copy it to the local scope
...
def test_...() -> None:
    factory._ENT_FACTORY = None
    # run the test knowing that the factory is now empty.
```

Otherwise this is a very brittle way of running these tests.

Reply:
Obligatory SO post: https://stackoverflow.com/questions/3536620/how-to-change-a-module-variable-from-another-module

Reply:
Oh actually we are doing something similar in an even easier way in other tests: https://github.com/getsentry/snuba/blob/master/tests/migrations/test_runner.py#L183

Reply:
Seeing something quite strange, it looks like `importlib.reload(factory)` does indeed reset `factory._DS_FACTORY` however, the threads are somehow still able to pick previous "working" states of the factory. My hypothesis for why this is happening is because `reload` only recompiles the code in the module on disk and re-evaluates it with the existing `sys.modules.__dict__`. As a result, a new factory module object is not being created which causes it to still referencing some old symbol (even though doing an `id()` on the old and reloaded factory return the same).

Reply:
I did but I think it makes the module a bit confusing to read. Also, the import method from the SO post requires all sub-modules of `snuba.datasets` to be implicitly imported in the test just to access the factory sub-module.

Instead I've revised the test cases to force a failure if the factory object initialization is decoupled from the mapping initialization. This seems to have fixed the variable name spacing issue and removes the need to spin up more threads (which is be inconsistent).